### PR TITLE
Init BindGen README

### PR DESCRIPTION
**Pull request by Spacesai1or**
Add a `README` for BindGen

### CODE REVIEW

**Comment by Spacesai1or** [comment]
Current dependencies on/for this PR:
* `develop`
  * **PR #8764** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8764?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8764?utm_source=stack-comment).

File: op-bindings/bindgen/README.md [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+
+All other default `"remote"` contract have some variation of the above execution flow depending on the nuances of each contract. For example:
+
+- `Create2Deployer`'s initialization and deployed bytecode is expected to differ from it's Optimism Mainnet deployment
+- `MultiSend_v130` has an `immutable` Solidity variable the resolves to `address(this)`, so we can't use the deployment bytecode from Ethereum Mainnet, we must get it's deployment bytecode from Optimism Mainnet
+- `SenderCreator` is deployed by `EntryPoint`, so it's initialization bytecode is provided in [artifacts.json](../artifacts.json) and not being fetched from Etherscan like other contracts
+
+#### Contracts that Don't Make Good Predeploys
+
+Not every contract can be added as a predeploy, and some contract's have nuances that make them potentially dangerous or troublesome to predeploy. Some examples of contracts that wouldn't make good predeploys:
```
Comment (mds1) [inline op-bindings/bindgen/README.md]:
```suggestion
Not every contract can be added as a predeploy, and some contracts have nuances that make them potentially dangerous or troublesome to predeploy. Below are some examples of contracts that wouldn't make good predeploys. This is not a comprehensive list, so make sure to use judgment for each contract added as a predeploy.
```

File: op-bindings/bindgen/README.md [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+
+- Contracts that haven't been audited or stood the test of time
+  - Once a contract is predeployed and a network is started, if a vulnerability is discovered for the contract and there is no way to disable the contract using some variable, the only option to "disable" the vulnerable contract is to either A. remove it from the L2 genesis and restart the L2 network B. Hardfork the network to remove/replace the predeploy or C. Warn users not to use the vulnerable predeploy
```
Comment (mds1) [inline op-bindings/bindgen/README.md]: nit to make the list more readable
```suggestion
  - Once a contract is predeployed and a network is started, if a vulnerability is discovered for the contract and there is no way to easily disable the contract, the only options to "disable" the vulnerable contract are to either (A) remove it from the L2 genesis and restart the L2 network, (B) Hardfork the network to remove/replace the predeploy, or (C) Warn users not to use the vulnerable predeploy
```

File: op-bindings/bindgen/README.md [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+- Related to above, contracts that may become deprecated/unsupported relatively soon
+  - As mentioned above, you're limited to options A, B, or C
+- Upgradeable Contracts
+  - While it's certainly feasible to predeploy an upgradeable contract, great care should be taken to ensure user's aren't "rugged" if the contract is upgraded to a malicious implementation. Understanding who has the ability to upgrade the contract is key to avoiding this. Additionally, user's might be expecting a predeploy to do something and may be caught off guard if the implementation was upgraded without their knowledge
```
Comment (mds1) [inline op-bindings/bindgen/README.md]: Replaced "rugged" with "security risks" wording, and added a new contract type
```suggestion
- Upgradeable Contracts
  - While it's certainly feasible to predeploy an upgradeable contract, great care should be taken to minimize security risks to users if the contract is upgraded to a malicious or buggy implementation. Understanding who has the ability to upgrade the contract is key to avoiding this. Additionally, user's might be expecting a predeploy to do something and may be caught off guard if the implementation was upgraded without their knowledge
- Contracts with Privileged Roles and Configuration Parameters
  - Similar to the upgradeable contracts, simply having an owner or other privileged role with the ability to make configuration changes can present a security risk and result in unexpected different behaviors across chains.
```

**Review by mds1 — COMMENTED** [review]
Read through the "Contracts that Don't Make Good Predeploys" section only—it looks great, just a few nits!

**Review by mds1 — APPROVED** [review]

File: op-bindings/bindgen/README.md [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+These flags are used with `all` and `remote` commands
+
+Flag                   | Type   | Description                                                                 | Required
+---------------------- | ------ | --------------------------------------------------------------------------- | --------
+`etherscan.apikey.eth` | String | An Etherscan API key for querying Ethereum Mainnet                          | Yes
+`etherscan.apikey.op`  | String | An Etherscan API key for querying Optimism Mainnet                          | Yes
+`rpc.url.eth`          | String | This is any HTTP URL that can be used to query an Ethereum Mainnet RPC node | Yes
+`rpc.url.op`           | String | This is any HTTP URL that can be used to query an Optimism Mainnet RPC node | Yes
+
+# Using BindGen to Add New Predeploys to L2 Genesis
```
Comment (annieke) [inline op-bindings/bindgen/README.md:162]: Can we add a small snippet here to specify that we're not looking to endorse more contracts at this moment? Something like 

> While we encourage hacking on the stack, we are not actively looking to integrate more contracts at an OP Stack's genesis.

File: op-bindings/bindgen/README.md [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+  - This has the same concern as differing initialization bytecode
+6. Lastly, the Go bindings are generated and the metadata file is written to the path provided as `metadata-out` CLI flag
+
+All other default `"remote"` contract have some variation of the above execution flow depending on the nuances of each contract. For example:
+
+- `Create2Deployer`'s initialization and deployed bytecode is expected to differ from it's Optimism Mainnet deployment
+- `MultiSend_v130` has an `immutable` Solidity variable the resolves to `address(this)`, so we can't use the deployment bytecode from Ethereum Mainnet, we must get it's deployment bytecode from Optimism Mainnet
+- `SenderCreator` is deployed by `EntryPoint`, so it's initialization bytecode is provided in [artifacts.json](../artifacts.json) and not being fetched from Etherscan like other contracts
+
+#### Contracts that Don't Make Good Predeploys
```
Comment (annieke) [inline op-bindings/bindgen/README.md:311]: this section is great!

**Review by annieke — COMMENTED** [review]
one small fix, other wise this looks great!

File: op-bindings/bindgen/README.md (hunk 4 above)
Comment (spacesailor24) [reply op-bindings/bindgen/README.md:162]: I added

```markdown
**Note** While we encourage hacking on the OP stack, we are not actively looking to integrate more contracts at the SuperChain's genesis.
```

Comment (Spacesai1or) [reply op-bindings/bindgen/README.md:162]: Updated to

```markdown
**Note** While we encourage hacking on the OP stack, we are not actively looking to integrate more contracts to the official OP stack genesis.
```

**Review by annieke — APPROVED** [review]

**Pull request merged**